Overhauling Hyperparameter Optimization

rbharath opened the pull request:
This PR pulls out another slice of #1792 that focuses on overhauling the hyperparameter optimization support. This PR isn't yet ready for review.

rbharath commented:
This PR is turning harder than i thought (seems to be a recurring theme while refactoring...). The reason for this is that we didn't previously have a cogent API for hyperparameter optimization. Gaussian Process Hyperparameter optimization and Grid search optimization followed completely different APIs, and there was no clear way to add a new hyperparameter optimization method. I think I have a handle on how to do this cleanly, but this PR is going to end up being a little bigger than I thought it would be. I'm hoping I can get this out and ready for review in the next few days. I'll add more details about the structural changes then.

nd-02110114 commented:
I have a question. Why does deepchem provide hyperparameter optimization APIs...?
I seem there are several good tools for optimizing hyper hyperparameter and deepchem don't need to provide them.
(but, I agree that it is better to provide tutorials about hyperparameter optimization and I will be able to help it)

Especially, I recommend Optuna. I think Optuna is the simplest and easiest tool. And, Optuna supports many optimization algorithm (RandomSearch, GridSerach, TPE, GP...) and many ML frameworks. (sklearn, TF, Keras, PyTorch....).

Optuna : https://github.com/optuna/optuna
Docs : https://optuna.readthedocs.io/en/stable/index.html
Sample : https://github.com/optuna/optuna#integrations

rbharath commented:
Good question! The hypeparameter API aims to provide a common API for wrappers around different hyperparameter frameworks. The Gaussian process model already uses pyGPGO for example. My hope is that by introducing a common API, we can create wrappers for other optimizers such as optuna, hyperopt, etc without much difficulty. 

The reason we need the wrappers is that these optimization frameworks aren't aware of DeepChem and don't provide out-of-the-box support for optimizing DeepChem models or handling DeepChem datasets. The wrappers themselves should be relatively simple to make once the common API is in place.

nd-02110114 commented:
Thanks you for comments! I could understand with your plan! This wrapper API is useful for someone.

But, I think deepchem should not focus on providing the wrapper for many optimizers. This is because it is hard to maintain this wrapper and there are some possibilities to lose some unique APIs which each optimizer has by making the common API. 

Currently, we try to add other DL framework (like PyTorch, JAX) models in deepchem, so it will make this wrapper more complicated and it will be hard for users to debug. I think many optimizers are easy to use compared with RDKit API or TF API and we should just provide only tutorials.

This opinion is just my own.... If you think deepchem should focus on providing the high-level API,  I think it is OK and please don't care my opinion.

rbharath commented:
I think you have a lot of good points! In an earlier issue https://github.com/deepchem/deepchem/issues/1754, I suggested something similar, that we should consider removing `dc.hyper` entirely.

After I wrote #1754 I talked to a lot of folks who've used DeepChem in various projects. One bit of feedback I got was that they found `dc.hyper` genuinely useful when getting started but found that it was limited as they became experts. I think the reason for this is that a lot of DeepChem's users are newcomers to machine learning who don't necessarily have a strong understanding of hyperparameter tuning. Having a sensible out-of-the-box implementation is useful for beginners I think.

At the same time, I think tutorials on how to optimize DeepChem models with optuna/hyperopt/etc are more useful for expert users. Experts won't necessarily want to use our simple hyperparameter optimizers. I propose that we do both. This PR just cleans up the Hyperparameter API and fixes the broken GaussianProcessOptimizer. After this is merged in, I'll open a new PR to track the addition of tutorials for optuna/hyperopt/etc. This way we can have an easy path for beginners but support for experts as well

nd-02110114 commented:
Thank you for sharing the details! 
After understanding the details, I think your direction is very reasonable 🙆‍♀️ .

rbharath commented:
Ok I think this PR is ready for review!

@peastman @nd-02110114 Could you do a review pass?

nd-02110114 commented on deepchem/hyper/gaussian_process.py:
```diff
@@ -5,140 +5,179 @@
 
+    # Demarcating internal function for readability
+    ########################
     def f(l00=0,
```
And, I thinks this fucntion is not readable....
I try to refactor, please check PR #1921

rbharath replied:
Merged in #1921. Thanks for the cleanup!

rbharath commented on deepchem/hyper/base_classes.py:
```diff
@@ -0,0 +1,93 @@
+
+  This class is an abstract base class for hyperparameter search
+  classes in DeepChem. Hyperparameter search is performed on
+  `dc.model.Model` classes. Each hyperparameter object accepts a
```
Good catch!

nd-02110114 replied:
Maybe you forget this point!

rbharath replied:
Good catch! Fixed

rbharath commented on deepchem/hyper/base_classes.py:
```diff
@@ -0,0 +1,93 @@
+
+    Parameters
+    ----------
+    model_class: constructor function.
```
Agreed! Will change over

rbharath commented on deepchem/hyper/base_classes.py:
```diff
@@ -0,0 +1,93 @@
+    >>> def rf_model_builder(model_params, model_dir):
+          sklearn_model = sklearn.ensemble.RandomForestRegressor(**model_params)
+          return dc.models.SklearnModel(sklearn_model, model_dir)
+
```
Good suggestion! Unfortunately `HyperparamOpt` is an abstract class that can't be initialized. Let me move the examples to the concrete subclasses instead though so we have an actual usage example.

rbharath commented on docs/hyper.rst:
```diff
@@ -8,20 +8,36 @@ learning algorithm used for the rest of learning and have to be set in
+single-threaded fashion. They are not intended to be production grade
+hyperparameter utilities, but rather useful first tools as you start
+exploring your parameter space. As the needs of your application grow,
+we recommend swapping to a more hyeavy duty hyperparameter
```
Fixed!

rbharath commented on scripts/install_deepchem_conda.sh:
```diff
@@ -36,4 +36,5 @@ conda install -y -q -c deepchem -c rdkit -c conda-forge -c omnia \
     pytest \
     pytest-cov \
     flaky
+yes | pip install pyGPGO
```
Good catch! Added in

rbharath commented:
I've made the changes @nd-02110114 asked for. I also added some more tests since I found a bug in the log handling for `GaussianProcessHyperparamOpt` and added an extra example.

@peastman @nd-02110114 I think this PR is ready for a second review

nd-02110114 commented on deepchem/hyper/tests/test_gaussian_hyperparam_opt.py:
```diff
@@ -0,0 +1,74 @@
+          max_iter=2)
+    valid_score = best_model.evaluate(self.valid_dataset, [metric],
+                                      transformers)
+    assert valid_score["pearson_r2_score"] > 0
```
Same here

peastman replied:
Those are distinct assertions that check different things.  Both of them seem useful.

rbharath replied:
Good suggestion! Added

nd-02110114 commented on deepchem/hyper/tests/test_gaussian_hyperparam_opt.py:
```diff
@@ -0,0 +1,74 @@
+
+    valid_score = best_model.evaluate(self.valid_dataset, [metric],
+                                      transformers)
+    assert valid_score["pearson_r2_score"] > 0
```
I think it is possible to assert more strictly like this.

```
assert valid_score["pearson_r2_score"] == max(all_results.values())
```

And, if you have some time to spare, it is better to write with use_max = False

rbharath replied:
Good suggestions! I've added the extra assertion and a test for minimization with `use_max=False`

nd-02110114 commented on deepchem/hyper/tests/test_grid_hyperparam_opt.py:
```diff
@@ -0,0 +1,157 @@
+from sklearn.ensemble import RandomForestRegressor
+
+
+class TestGridHyperparamOpt(unittest.TestCase):
```
I think it is better to add assertion about valid_score like GPGO test in all tests.

```
assert valid_score["your_metric"] == max(all_results.values())
```

rbharath replied:
Added in!

peastman commented on deepchem/hyper/grid_search.py:
```diff
@@ -9,23 +7,39 @@
-    self.verbose = verbose
+  >>> import sklearn
+  >>> import deepchem as dc
+  >>> def rf_model_builder(**model_params):
+  ...   rf_params = {k: v for (k, v) in model_params.items() if k != 'model_dir'}
+  ...   model_dir = model_params['model_dir']
+  ...   sklearn_model = sklearn.ensemble.RandomForestRegressor(**rf_params)
+  ...   return dc.models.SklearnModel(sklearn_model, model_dir)
```
This may not be the best example.  For any standard DeepChem model the code would be much simpler, right?

```python
def gc_model_builder(**model_params):
  return dc.models.GraphConvModel(**model_params)
optimizer = dc.hyper.GridHyperparamOpt(gc_model_builder)
```

Or you could even skip defining the function and just write

```python
optimizer = dc.hyper.GridHyperparamOpt(lambda **p: dc.models.GraphConvModel(**p))
```

rbharath replied:
This is a great suggestion! Much cleaner. Swapped over

peastman commented on deepchem/hyper/tests/test_grid_hyperparam_opt.py:
```diff
@@ -0,0 +1,157 @@
+    w_train = np.ones_like(y_train)
+    ids_train = ["C"] * n_train
+
+    train_dataset = dc.data.DiskDataset.from_numpy(X_train, y_train, w_train,
+                                                   ids_train, tasks)
```
Why write out the two datasets to disk?  Why not just create a NumpyDataset?

rbharath replied:
Good suggestion! Fixed

peastman commented on deepchem/hyper/tests/test_grid_hyperparam_opt.py:
```diff
@@ -0,0 +1,157 @@
+    params_dict = {"layer_sizes": [(10,), (100,)]}
+
+    def model_builder(**model_params):
+      model_dir = model_params['model_dir']
+      multitask_params = {
+          k: v for (k, v) in model_params.items() if k != 'model_dir'
+      }
```
I don't think there's any need to treat `model_dir` separately.

rbharath replied:
Reworked the tests so this example isn't there anymore.

peastman commented on examples/hyperparam_opt/gaussian_hyperparam_opt.py:
```diff
@@ -0,0 +1,26 @@
+metric = dc.metrics.Metric(dc.metrics.pearson_r2_score)
+
+
+def rf_model_builder(**model_params):
+  rf_params = {k: v for (k, v) in model_params.items() if k != 'model_dir'}
+  model_dir = model_params['model_dir']
+  sklearn_model = sklearn.ensemble.RandomForestRegressor(**rf_params)
+  return dc.models.SklearnModel(sklearn_model, model_dir)
```
Here again this isn't a good example.  It will lead to people thinking their code has to be more complicated than it really does.  Let's use a standard DeepChem model for the examples.

rbharath replied:
Swapped out for a simpler example

peastman commented on deepchem/hyper/base_classes.py:
```diff
@@ -0,0 +1,83 @@
+    Parameters
+    ----------
+    params_dict: dict
+      Dictionary mapping strings to Ints/Floats/Lists. Note that the
```
Are those really the only allowed types?  What if a model takes arguments of a different type, like a string or a bool?  It needs to explain the significance of those types and how this parameter is interpreted.

rbharath replied:
Good point! I've fixed this and added some more clarifying documentation

peastman commented on deepchem/hyper/gaussian_process.py:
```diff
@@ -5,238 +5,249 @@
+  for key in keys:
+    filename += "_%s" % str(key)
+    value = hyper_params[key]
+    if isinstance(value, int):
+      filename += "_%s" % str(value)
+    else:
+      filename += "_%.2f" % value
```
If `value` is a list this will throw an exception.

rbharath replied:
Good point! Fixed

peastman commented on deepchem/hyper/gaussian_process.py:
```diff
@@ -5,238 +5,249 @@
+  models. If you don't have pyGPGO installed, you won't be able to use
+  this class.
+
+  Note that `params_dict` has a different semantics than for
+  `GridHyperparamOpt`. `param_dict[hp]` must be an int/float and is
+  used as the center of a search range.
```
The documentation in this file is inconsistent.  Sometimes it says int/float/list, and other times only int/float.  I think the latter is the correct one?  Explain this in more detail: GPGO can only optimize numerical hyperparameters, so all others have to be fixed.  And give an example showing how to specify both the fixed and variable hyperparameters.  For example,

 ```python
def model_builder(**model_params):
  n_layers = model_params['layers']
  layer_width = model_params['width']
  dropout = model_params['dropout']
  return dc.models.MultitaskClassifier(
    n_tasks=5,
    n_features=100,
    layer_sizes=[layer_width]*n_layers,
    dropouts=dropout
  )
```

On the other hand, I'm not sure this is the best design.  It would be simpler for users if they could specify arbitrary parameters of arbitrary type.  Perhaps `hyperparam_search` should let you specify two dicts: one for parameters to optimize, and one for parameters that are just passed through unchanged?  Or there could be an optional argument taking the names of the parameters to optimize (with the default being to optimize everythin)?  Or maybe `GaussianProcessHyperparamOpt` should permit you to include lists, but require they must have length 1 (only one possible value)?

rbharath replied:
I did a bit of thinking and I think the current design with a custom `model_builder` is cleanest. This design matches what I've seen in other hyperparameter optimization libraries and allows for user flexibility in controlling arbitrary arguments to the model. I've added in the example you've suggested.

I've also fixed the documentation to only read int/float.

peastman commented on deepchem/hyper/gaussian_process.py:
```diff
@@ -5,238 +5,249 @@
+    max_iter: int, (default 20)
       number of optimization trials
-    search_range: int(float)
+    search_range: int(float) (default 4)
```
Is it reasonable to assume this will always have the same value for all hyperparameters?

rbharath replied:
Good point. I've fixed it so that it can accept different search ranges for each parameter

rbharath commented:
I've addressed most of the review comments but have a few more that I'll work through tomorrow.

rbharath commented:
I think this PR is almost done with one issue remaining. I'm getting some strange checkpoint warnings from TensorFlow with I optimize TensorFlow models. For example:

```
WARNING:tensorflow:Inconsistent references when loading the checkpoint into this object graph. Either the Trackable object references in the Py
thon program have changed in an incompatible way, or the checkpoint was generated in an incompatible program.
                                                                                                                                               
Two checkpoint references resolved to different objects (<tf.Variable 'private__graph_conv_keras_model_5/graph_conv_10/kernel:0' shape=(75, 64)
 dtype=float32, numpy=                                                                                                                         array([[-0.00611123, -0.02878498, -0.12021131, ...,  0.17456074,                                                                               
        -0.15475217, -0.18777789],                                                                                                             
       [-0.1489263 ,  0.03313252,  0.08405624, ..., -0.18309508,                                                                               
         0.17885186, -0.03379297],                                                                                                             
       [-0.08297385,  0.14736705,  0.03052345, ..., -0.10705137,                                                                               
        -0.09505855, -0.02651718],                                     
       ...,                                                            
       [-0.09474242, -0.09837816,  0.02341484, ..., -0.14325964,                                                                               
         0.13704158,  0.20472397],
       [ 0.15134431, -0.18757658,  0.13552605, ...,  0.20161061,
         0.16667165, -0.07482423],
       [-0.06389859, -0.14159493,  0.12783583, ..., -0.05287665,
        -0.05681728,  0.03825043]], dtype=float32)> and <tf.Variable 'private__graph_conv_keras_model_5/graph_conv_10/kernel:0' shape=(75, 64) 
dtype=float32, numpy=
array([[-0.19917306,  0.00908746, -0.05600289, ..., -0.1040943 ,
        -0.00210239,  0.04228573],
       [-0.00108278,  0.1207066 , -0.20167896, ...,  0.20652021,
        -0.15815929, -0.12034214],
       [-0.17928252, -0.0565967 ,  0.07985629, ..., -0.05842453,
        -0.10541276, -0.17301702],
       ...,
       [ 0.11975063,  0.13149019, -0.13186957, ...,  0.05369894,
        -0.18793763, -0.0703835 ],
       [ 0.19162537,  0.2029113 ,  0.1383634 , ...,  0.00662975,
        -0.09817264,  0.1355009 ],
       [ 0.12863006, -0.13623261, -0.1724247 , ...,  0.002299  ,
        -0.14834997, -0.13962182]], dtype=float32)>).
```

I'm not sure what could be causing these warnings right now.

@peastman Have you seen similar warnings before?

rbharath commented:
As on additional point, while making revisions, I noticed we have an inconsistency in our model reloading functionality:

- `KerasModel.restore()`
- `Model.reload()`, `SklearnModel.reload()`

This came up in the current PR since we have to reload trained models to return the strongest model after HP tuning. I suggest we use `restore()` as the default across all models since `KerasModel` models are our most used models and might be the default that users are used to already.

rbharath commented:
I think some of the statements I was seeing came from https://stackoverflow.com/questions/58289342/tf2-0-translation-model-error-when-restoring-the-saved-model-unresolved-objec, while others are from the bug https://github.com/deepchem/deepchem/issues/1943. I think #1943 isn't related to this PR so I think once I've finished cleaning up the PR (need to fix up some tests), this will be ready for review (and hopefully merge). We can fix #1943 on a separate PR I think

rbharath commented:
As one additional comment, I realized that the changes in this PR would break `dc.molnet.run_benchmark`. We don't have any tests for these methods which makes them additionally easy to break. I'll fix the breakage and see if I can add a basic unit test.

rbharath commented:
I've factored out the needed moleculenet fixes in #1950. I also refactored `GaussianProcessHyperparamOpt` to avoid #1943. I also removed the change to `Model.restore()` from this PR and raised a separate issue #1951 to discuss. At this point, if tests pass, I think this PR is ready to merge.

rbharath commented:
I'm going to go ahead and merge in this PR. 

@peastman @nd-02110114 Please feel free to give any feedback or comments you have; I'll fix issues in #1950
